### DISCUSSION TITLE
Pico: How to send/receive USB Gadget MIDI Sysex Traffic via Network?

### DISCUSSION THREAD

**Discussion by ajackson99** [tttapa/Control-Surface#1018] [Q&A]
Hey,

my requirements:

- set a custom Product Descriptor for the USB MIDI Gadget
- set a custom Cable name for it
- forward the received traffic via Network (AppleMIDI, UDP, anyhow) via Network to a host
- beeing able to send sysex from host to Pico, which forwards it to its USB MIDI connected device

The detailed description:
Pico is connected to an iPad via CamKit - an App interacts only if if the ProductDescriptor is f.Ex "JohnDoe Interface" with Portname "Port C" (CoreMIDI)
Now, i would like to be able to interact with the Pico from a remote host. That means send and receive Sysex Messages.

The problem i have, it seems not be possible to set custom descriptor-/cable names in Controls Surface.
With TinyUSB its possible - but the pipes seem not to be compatible with TinyUSB. 

F.ex:
TinyUSBDevice.setProductDescriptor("John Doe Inteface");
usb_midi.setCableName(1, "Port C");

Is there any easy possibility to achieve this with Control Surface?

If yes, how to setup a pipe which mirrors the traffic via Network?

Tried a lot but no success till now. 

Any hints?

Thanks in advance.

**Comment by tttapa** [comment]
> The problem i have, it seems not be possible to set custom descriptor-/cable names in Controls Surface.

This depends on the MIDI over USB backend used. USB descriptors are often configured by the Arduino core or by other libraries.

If you're using a Pi Pico with the Adafruit TinyUSB backend, you can use the following:
```cpp
#include <Control_Surface.h>

USBMIDI_Interface midi;

void setup() {
  TinyUSBDevice.setManufacturerDescriptor("ajackson99");
  TinyUSBDevice.setProductDescriptor("John Doe Inteface");
  midi.backend.backend.setCableName(1, "Port C");
  midi.begin();
  // ...
}
```
(`midi.backend.backend` is an instance of the `Adafruit_USBD_MIDI` class.)

> With TinyUSB its possible - but the pipes seem not to be compatible with TinyUSB.

What do you mean?

> If yes, how to setup a pipe which mirrors the traffic via Network?

You can find more information here:
- https://tttapa.github.io/Control-Surface-doc/Doxygen/d3/df7/midi-tutorial.html#midi_md-routing
- https://tttapa.github.io/Control-Surface-doc/Doxygen/dc/d7d/AppleMIDI_8ino-example.html

The `AppleMIDI_interface` variable in that example is a Control Surface MIDI interface that can be used with pipes.

**Comment by ajackson99** [comment]
Thank you very much for your hint! 

This was the missing point by me:

> (midi.backend.backend is an instance of the Adafruit_USBD_MIDI class.)

Will try this out soon and report back.


> What do you mean?
> 
> If yes, how to setup a pipe which mirrors the traffic via Network?

Yes, i noticed this before and was able to build the AppleMIDI example - works fine.
Sorry, maybe i was not clear enough. Almost afraid to ask:

I meant if there exists an example how to setup both MIDI interfaces on the pico w:
- One via USB (TinyUSBDevice)
- The second via AppleMIDI
- And the pipe them, based on the pipe examples via Control Surface, as its in the docs: https://tttapa.github.io/Control-Surface-doc/Doxygen/dc/d7d/AppleMIDI_8ino-example.html
- I mean: There are examples between serial- and usb- MIDI. Ok so far... So this should also work between AppleMIDI_interface and an USBMIDI_Interface. Right?
